tests/failpoints/cases/test_unsafe_recovery.rs

```diff
@@ -92,3 +86,71 @@ fn test_unsafe_recover_send_report() {
     assert_eq!(reported, true);
     fail::remove("on_handle_apply_store_1");
 }
+
+#[allow(clippy::mutex_atomic)]
+#[test]
+fn test_unsafe_recover_wait_for_snapshot_apply() {
+    let mut cluster = new_server_cluster(0, 3);
+    cluster.cfg.raft_store.raft_log_gc_count_limit = 8;
+    cluster.cfg.raft_store.merge_max_log_gap = 3;
+    cluster.cfg.raft_store.raft_log_gc_tick_interval = ReadableDuration::millis(10);
+    cluster.run();
+    let nodes = Vec::from_iter(cluster.get_node_ids());
+    assert_eq!(nodes.len(), 3);
+
+    let pd_client = Arc::clone(&cluster.pd_client);
+    pd_client.disable_default_operator();
+    let region = block_on(pd_client.get_region_by_id(1)).unwrap().unwrap();
+    configure_for_lease_read(&mut cluster, None, None);
+
+    // Makes the leadership definite.
+    let store2_peer = find_peer(&region, nodes[1]).unwrap().to_owned();
+    cluster.must_transfer_leader(region.get_id(), store2_peer);
+    cluster.stop_node(nodes[1]);
+    // at least 4m data
+    (0..10).for_each(|_| cluster.must_put(b"random_k", b"random_v"));
+    // Sleep for a while to ensure all logs are compacted.
+    sleep_ms(100);
+    // Makes the group lose its quorum.
+    cluster.stop_node(nodes[2]);
+
+    // Blocks the raft snap apply process.
+    let apply_triggered_pair = Arc::new((Mutex::new(false), Condvar::new()));
```

Comment:
Why not use channel?

Reply:
`apply_triggered` is a multi-productions-one-consumption model, while `apply_released` is a one-production-multi-consumptions model. So, I think channel is not best option for both models here? Since the channel may be filled up or over consumed?

Reply:
You mean:
```
let (triggered_tx, triggered_rx) = mpsc::bounded(1);
let (released_tx, released_rx) = mpsc::bounded(1);
thread::spawn(move || {
let _ = triggered_tx.send(());
let _ = released_rx.recv(());
});

triggered_rx.recv_timeout().unwrap();
drop(release_tx);
```
?

Reply:
Done